src/main/java/com/redhat/devtools/lsp4ij/features/navigation/LSPGotoDeclarationHandler.java

```diff
@@ -42,25 +52,75 @@ public class LSPGotoDeclarationHandler implements GotoDeclarationHandler {
 
     @Nullable
     @Override
-    public PsiElement[] getGotoDeclarationTargets(@Nullable PsiElement sourceElement, int offset, Editor editor) {
+    public PsiElement[] getGotoDeclarationTargets(@Nullable PsiElement sourceElement,
+                                                  int offset,
+                                                  Editor editor) {
         Project project = editor.getProject();
         if (project == null || project.isDisposed()) {
             return PsiElement.EMPTY_ARRAY;
         }
-        PsiFile psiFile = sourceElement.getContainingFile();
+        PsiFile psiFile = sourceElement != null ? sourceElement.getContainingFile() : null;
         if (psiFile == null) {
             return PsiElement.EMPTY_ARRAY;
         }
         if (!LanguageServersRegistry.getInstance().isFileSupported(psiFile)) {
             return PsiElement.EMPTY_ARRAY;
         }
+
+        // If this was called for a populated semantic tokens view provider, try to get the target directly from it
+        LSPSemanticTokensFileViewProvider semanticTokensFileViewProvider = LSPSemanticTokensFileViewProvider.getInstance(psiFile);
+        if (semanticTokensFileViewProvider != null) {
+            // If this is definitely a reference, resolve it
+            if (semanticTokensFileViewProvider.isReference(offset)) {
```

Comment:
If we can respond definitively based on semantic token information saying this is a reference or a declartion, do so.